geps/gep-2648/index.md

```diff
@@ -0,0 +1,412 @@
+# GEP-2648: Direct Policy Attachment
+
+* Issue: [#2648](https://github.com/kubernetes-sigs/gateway-api/issues/2648)
+* Status: Experimental
+
+(See status definitions [here](/geps/overview/#gep-states)
+
+## TLDR
+
+Describe and specify a design pattern for a class of metaresource that can
+affect specific settings across a single target object.
+
+This is a design for a _pattern_, not an API field or new object.
+
+## Goals
+
+* Specify what common properties all Direct Attached Policies MUST have
+* Recommend design patterns for areas that cannot be mandated but that could
+  cause problems for API designers.
+
+## Non-Goals
+
+* Fully specify the entire design space for Direct Policy Attachment
+
+## Introduction
+
+GEP-713 defines two classes of Policy Attachment: Direct and Inherited.
+
+Direct Attached Policies (or Direct Policies) _only_ affect the object they are
+attached to; that is, the object specified in their `targetRef`.
+
+Note that as soon as the Policy affects more objects than the referenced object,
+it is an Inherited Policy.
+
+## Direct Policy Attachment requirements in brief
+
+The following parts of GEP-713 also apply here. Direct Policy Attachments:
+
+- MUST be their own CRDs (e.g. `TimeoutPolicy`, `RetryPolicy` etc),
+- MUST include both `spec` and `status` stanzas
+- MUST have the `status` stanza include a `conditions` section using the standard
+  upstream Condition type
+- MUST use the `targetRef` struct to specify their target resource.
+- MUST follow the naming requirements (MUST be named to clearly indicate that the
+kind is a Policy, and SHOULD use the `Policy` suffix at the end of the Kind and
+`policies` at the end of the Resource names).
+- MAY target a subsection of a resource using the `sectionName` field of the
+  `targetRef` struct. If it does, it MUST abide by the rules in this document,
+  and MUST clearly indicate the objects that `sectionName` may be used on in its
+  documentation.
+
+## Direct Policy Attachment
+
+A Direct Policy Attachment is tightly bound to one instance of a particular
+Kind within a single namespace (or to an instance of a single Kind at cluster scope),
+and only modifies the behavior of the object that matches its binding.
+
+As an example, the BackendTLSPolicy is specified in GEP-1897, BackendTLSPolicy -
+Explicit Backend TLS Connection Configuration. This Policy attaches to the
+Service object and tells Gateway API implementations what TLS settings should be
+used to connect to that Service when it is used as a backend by a Route.
+
+See GEP-1897 for all the details of this Policy object.
+
+## Direct Policy design guidelines
+
+With this example in mind, here are some guidelines for when to consider
+using Direct Policy Attachment:
+
+* The number or scope of objects is exactly _one_ object.
+* The modifications to be made to the objects don’t have any transitive information -
```

Comment:
I don't think its clear what it means to "only affect the single object". I don't feel like there are any definitions for what the scope of such objects are.

To make things more concrete, what if I apply a BackendTimeoutPolicy to a Gateway. What are the implications here?

Reply:
If the backend timeouts are set per-backend, then it's an Inherited Policy, because there's no backend config in Gateway.

The intent here is to tightly constrain what a Direct Attached Policy is so that we don't have to boil the ocean of finding a good way to tell everyone whose objects are affected by a policy about that in the status of all the resources.

It's also about ensuring that we don't have "Direct Attached" Policy objects that effectively set defaults, even if the `defaults` stanza isn't used.

Reply:
I added some more text here.

Reply:
The goals make sense,I am just trying to fully grok the meaning of this since it seems like this is an important aspect 

What does it mean to have a timeout set per-backend?

I can say logically that a timeout seems to apply to a backend in my mental model. And maybe in proxy implementations. But I could have an implementation that has one global value "timeout", and Gateway is the only logical place to put that policy in that implementation. Is that ok, or does that violate this?

What if my data plane implementation does happen to have per-backend config, but I don't want to deal with it so I only support gateway level config - is that ok?

More broadly, what exactly defined what is the scope of impact of an object? Is it the ability to attach a single type to multiple places? Is it the implementation details of how a policy is implemented? Is it a users mental model of how it works?

Reply:
I think this question overlaps with the session persistence GEP @gcs278 has been working on (#2634). In that case we've found that some implementations (Envoy or GRPC-based in particular) can only attach this config at Route level. So even if a user attaches the config at the Backend layer, the underlying implementation is going to be implementing it at the Routing layer. Would that go against this guideline?

Similarly, that GEP currently proposes that Routes will include inline overrides that could override the values specified on a Service. That's not really policy at that point, but it is some config overriding a direct-attached policy which I believe would be a new concept for us. 

To be clear, I actually like the approach that GEP is proposing, I'm just trying to make sure that the corresponding changes we're making here can line up with that so we don't end up with conflicting GEPs.

Reply:
Good questions, thanks @howardjohn

> I can say logically that a timeout seems to apply to a backend in my mental model. And maybe in proxy implementations. But I could have an implementation that has one global value "timeout", and Gateway is the only logical place to put that policy in that implementation. Is that ok, or does that violate this?

I guess in this case it would be a Direct Attached Policy, yes. It's all about _which resources you need to know about to understand the state of the system_, at the end of the day.
 
> What if my data plane implementation does happen to have per-backend config, but I don't want to deal with it so I only support gateway level config - is that ok?

If you only need to know about the Policy and the Gateway to know everything you need to, then yes.

> More broadly, what exactly defined what is the scope of impact of an object? Is it the ability to attach a single type to multiple places? Is it the implementation details of how a policy is implemented? Is it a users mental model of how it works?

The best way I've found to think of the scope is which objects are _impacted_ by the Policy. That's going to depend on exactly what the Policy actually _does_, and so is going to be Policy-specific.

To use your examples:
- being able to attach a single type to multiple places is almost certainly going to end up as an Inherited Policy, because _probably_, the settings will effectively be defaults or overrides (if they can be attached at different levels). If a Policy can only be attached to sibling objects in a hierarchy (only to Routes, for example), _and_ only affects properties of those sibling objects (so, attaching to Routes can't change backend properties for Service), then it can be a Direct Attached Policy.
- Is the implementation details of how a policy is implemented? In a word, yes. Policy writers will need to be specific about describing what a Policy affects.
- Is it a user's mental model of how it works? Ideally, this model should be pretty similar if not identical to how it's implemented, and the Policy should be able to _clearly indicate_ what objects it is affecting. What "affecting" means depends on the settings defined by the Policy and the implementation.

My personal guideline is "If I'm a user, how will I know if my thing is affected by a Policy?" If there's _any_ level of action-at-a-distance, then the Policy is an Inherited one.

If you have suggestions on better ways to communicate this, happy to change things up more.

Reply:
> If there's any level of action-at-a-distance, then the Policy is an Inherited one.

I think there's probably a category of policies that we know will exclusively be attached at one level and thus would never benefit from defaults and overrides, but would also still have some knock-on effects above or below where they're attached. 

It may be simpler to just say "direct policies can be attached at exactly one layer while inherited policies are used when a policy can be attached at more than one layer with support for defaults and overrides". I think that may still achieve the goal you're aiming for here while also leaving room for these inevitable policies that clearly only fit in one place but could indirectly affect other layers.

Reply:
Do we need take in policies that are additive? They don't make sense to have defaults and overrides, but do target multiple layers. Authorization policies are likely to fall info this category

Reply:
🤔 that's a good point. Are you referring to policies that essentially just have implicit defaulting behavior all the way down the stack? For example:

policy attached to GW:
```
foo: 1
bar: 2
```

policy attached to Route:
```
bar: 4
baz : 3
```

computed result:
```
foo: 1
bar: 4
baz: 3
```

Reply:
I was thinking:
GatewaY:
```
rules: [a]
```

Route:
```
rules: [b]
```

Result: `rules: [a,b]`

Reply:
At the end of the day, what I'm trying to do is make it clear that Direct Policies have an easier time with status. If you do _any_ cross object inheritance as mentioned in the previous two comments, those count as inherited, _because you have to solve the Inherited status problem_ to have full status.

"Direct Policies can be attached at exactly one layer" is a good _start_, but if they have knock-on effects further down the hierarchy, then again, _they have the same status problem as Inherited Policies_.

Maybe I just make it more explicit here:
- Direct Policies can be attached at exactly one layer
- Direct Policies can have effects only on the layer they attach to
- Direct Policies can effects only on the object withing the layer that they attach to

This stops the "but my Direct Policy affected another Route attached to the same Gateway when attached to a Route" problem which could arise if we only had the first two items.

Reply:
Oh, and @howardjohn, additive policies being Inherited is _definitely_ intended, that's one of the things I was worried that the original definition missed. That kind of implicit default still has all the status problems of a more explicit Inherited policy with a defined `defaults` section, so I wanted to make sure that we caught that in the definitions.